spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/core/GcpContextAutoConfiguration.java

```diff
@@ -72,4 +75,11 @@ public GcpProjectIdProvider gcpProjectIdProvider() {
 
 		return projectIdProvider;
 	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	@ConditionalOnProperty("VCAP_SERVICES")
+	public PcfConfiguration pcfConfiguration() {
+		return new DefaultPcfConfiguration(System.getenv("VCAP_SERVICES"));
```

Comment:
I would suggest you to pay attention to the `CloudFoundryVcapEnvironmentPostProcessor` in Spring Boot and don't worry about this ENV var at all.

I think there should be something in the [Reference Manual](https://docs.spring.io/spring-boot/docs/2.0.0.RC2/reference/htmlsingle/#production-ready-cloudfoundry) as well.

Reply:
After looking at `CloudFoundryVcapEnvironmentPostProcessor`, I don't think we can use it because it contains user-specified values as key.

In other words, say I create a Pub/Sub service (with `cf create-service google-pubsub standard cactuar`). `cactuar` is the service name that I chose. The resulting JSON will be:

```
"google-pubsub": [
   {
    "credentials": {
     "Email": "pcf-binding-3352ec74@graphite-test-spring-cloud-gcp.iam.gserviceaccount.com",
     "Name": "pcf-binding-3352ec74",
     "PrivateKeyData": "**ellided**",
     "ProjectId": "graphite-test-spring-cloud-gcp",
     "UniqueId": "118335157430758507928",
     "subscription_name": "",
     "topic_name": "pcf_sb_4_1521138145589148714"
    },
    "label": "google-pubsub",
    "name": "mypubsub",
    "plan": "default",
    "provider": null,
    "syslog_drain_url": null,
    "tags": [
     "gcp",
     "pubsub"
    ],
    "volume_mounts": []
   }
  ]
```

`CloudFoundryVcapEnvironmentPostProcessor` then transforms this into properties with the prefix `vcap.services.cactuar`. For example, `vcap.services.cactuar.credentials.PrivateKeyData`.

Alternatives to this are welcome!

Reply:
If I were you I'd look at binding `vcap.services.*` to a `Map` and pull the data out of there. Did you try that? Either that or use a naming convention for the service instances (and make it configurable of course). Also see here for another approach: https://github.com/spring-cloud/spring-cloud-cloudfoundry/blob/master/spring-cloud-cloudfoundry-web/src/main/java/org/springframework/cloud/cloudfoundry/environment/VcapServiceCredentialsEnvironmentPostProcessor.java

Reply:
@dsyer not sure if I'm missing anything. As I mentioned, `vcap.services.*` are indexed by user-specified keys. There is no way we can know what the user will pick. In the example I gave, I called my Pub/Sub service `cactuar` and that resulted in `vcap.services.cactuar`.

I understand that needs to be because the same service can be bound several times to the same application. In that scenario, we're actually failing, because we don't have any criteria to choose from any of the service accounts.

Also, as I understand, post processors kick in after then context is built, whereas we need the information from `VCAP_SERVICES` to build the context.

Let me know if I'm missing anything or if there are other alternatives. For now, this seems to be the only one that works for our requirements.